searx/external_bang.py

```diff
@@ -1,14 +1,73 @@
-from searx.data import bangs_loader
+# SPDX-License-Identifier: AGPL-3.0-or-later
 
-# bangs data coming from the following url convert to json with
-# https://raw.githubusercontent.com/jivesearch/jivesearch/master/bangs/bangs.toml
-# https://pseitz.github.io/toml-to-json-online-converter/
-# NOTE only use the get_bang_url
+from searx.data import EXTERNAL_BANGS
 
-bangs_data = {}
-for bang in bangs_loader()['bang']:
-    for trigger in bang["triggers"]:
-        bangs_data[trigger] = {x: y for x, y in bang.items() if x != "triggers"}
+
+def get_node(bang):
+    node = EXTERNAL_BANGS['trie']
+    after = ''
+    before = ''
+    for bang_letter in bang:
+        after += bang_letter
+        if after in node:
+            node = node[after]
+            before += after
+            after = ''
+    return node, before, after
+
+
+def get_bang_definition_and_ac(bang):
+    node, before, after = get_node(bang)
+
+    bang_definition = None
+    bang_ac_list = []
+    if after != '':
+        for k in node:
+            if k.startswith(after):
+                bang_ac_list.append(before + k)
+    elif isinstance(node, dict):
+        bang_definition = node.get('*')
+        bang_ac_list = [before + k for k in node.keys() if k != '*']
+    elif isinstance(node, str):
+        bang_definition = node
+        bang_ac_list = []
+
+    return bang_definition, bang_ac_list
+
+
+def resolve_bang_definition(bang_definition, query):
+    result = bang_definition.split(chr(1))
+    result[0] = result[0].replace(chr(2), query)
+    if result[0].startswith('//'):
+        result[0] = 'https:' + result[0]
+    result[1] = int(result[1]) if len(result[1]) > 0 else 0
+    return result
+
+
+def get_bang_definition_and_autocomplete(bang):
+    bang_definition, bang_ac_list = get_bang_definition_and_ac(bang)
+
+    new_autocomplete = []
+
+    current = [*bang_ac_list]
+    done = set()
+    while len(current) > 0:
+        bang_ac = current[0]
+        del current[0]
+        done.add(bang_ac)
+
+        bang_definition, bang_ac_list = get_bang_definition_and_ac(bang_ac)
```

Comment:
Nit: `bang_definition` and `bang_ac_list` shadows the variable from the outer scope. I would prefer that you renamed these to `current_bang_definition` and `current_bang_ac_list`. This way the code becomes more readable and we do not accidentally make mistakes if/when these variables are renamed.